File: events/src/main/java/com/nerdscorner/mvplib/events/behaviour/MvpEventsBehaviour.java
package com.nerdscorner.mvplib.events.behaviour;

import org.greenrobot.eventbus.EventBus;

import android.content.Intent;
import android.content.res.Configuration;
import android.os.Bundle;
import android.os.Parcel;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.MenuItem;
import com.nerdscorner.mvplib.commons.behaviour.Behaviour;
import com.nerdscorner.mvplib.commons.mvp.presenter.BasePresenter;

public class MvpEventsBehaviour extends Behaviour {

    private final static int CONTENTS_DEFAULT = 0;

    protected BasePresenter presenter;

    public final static Creator<MvpEventsBehaviour> CREATOR = new Creator<MvpEventsBehaviour>() {
        @Override
        public MvpEventsBehaviour createFromParcel(final Parcel in) {
            return new MvpEventsBehaviour(in);
        }

        @Override
        public MvpEventsBehaviour[] newArray(final int size) {
            return new MvpEventsBehaviour[size];
        }
    };

    public MvpEventsBehaviour(BasePresenter presenter) {
        this.presenter = presenter;
    }

    /**
     * Default scope because it's used from {@link #CREATOR}.
     *
     * @param in The parcel with this type values.
     */
    /* default */ MvpEventsBehaviour(final Parcel in) {
        super(in);
    }

    @Override
    public int describeContents() {
        return CONTENTS_DEFAULT;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
    }

    @Override
    public void onStart() {
        super.onStart();
        presenter.onStart();
    }

    @Override
    public void onResume() {
        super.onResume();
        try {
            EventBus.getDefault().register(presenter);
        } catch (Exception ignored) {
            //No @Subscribe annotations detected
        }
        presenter.onResume();
    }

    @Override
    public void onPause() {
        super.onPause();
        presenter.onPause();
        try {
            EventBus.getDefault().unregister(presenter);
        } catch (Exception ignored) {
            //No @Subscribe annotations detected
        }
    }

    @Override
    public void onStop() {
        super.onStop();
        presenter.onStop();
    }

    @Override
    public boolean onBackPressed() {
        return !presenter.onBackPressed() && super.onBackPressed();
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        return !presenter.onOptionsItemSelected(item) && super.onOptionsItemSelected(item);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        presenter.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        presenter.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    @Override
    public void onConfigurationChanged(@NonNull Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        presenter.onConfigurationChanged(newConfig);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        presenter.onSaveInstanceState(outState);
    }

    @Override
    public void onRestoreInstanceState(@NonNull Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        presenter.onRestoreInstanceState(savedInstanceState);
    }

    @SuppressWarnings("unchecked")
    @Nullable
    @Override
    public <Component> Component getComponent(@NonNull Class<Component> componentClass) {
        if (componentClass.isAssignableFrom(EventBus.class)) {
            return (Component) EventBus.getDefault();
        }
        return null;
    }
}

File: events/src/main/java/com/nerdscorner/mvplib/events/behaviour/MvpEventsBehaviour.kt
package com.nerdscorner.mvplib.events.behaviour

import android.content.Intent
import android.content.res.Configuration
import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import com.nerdscorner.mvplib.events.presenter.BasePresenter
import kotlinx.android.parcel.Parcelize
import kotlinx.android.parcel.RawValue
import org.greenrobot.eventbus.EventBus

@Parcelize
class MvpEventsBehaviour(val presenter: @RawValue BasePresenter<*, *>) : Behaviour() {

    override fun onStart() {
        super.onStart()
        presenter.onStart()
    }

    override fun onResume() {
        super.onResume()
        try {
            EventBus.getDefault().register(presenter)
        } catch (ignored: Exception) {
            //No @Subscribe annotations detected
        }

        presenter.onResume()
    }

    override fun onPause() {
        super.onPause()
        presenter.onPause()
        try {
            EventBus.getDefault().unregister(presenter)
        } catch (ignored: Exception) {
            //No @Subscribe annotations detected
        }

    }

    override fun onStop() {
        super.onStop()
        presenter.onStop()
    }

    override fun onBackPressed(): Boolean {
        return !presenter.onBackPressed() && super.onBackPressed()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        return presenter.onCreateOptionsMenu(menu)
    }

    override fun onCreateOptionsMenu(menu: Menu?, inflater: MenuInflater?) {
        presenter.onCreateOptionsMenu(menu, inflater)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return !presenter.onOptionsItemSelected(item) && super.onOptionsItemSelected(item)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        presenter.onActivityResult(requestCode, resultCode, data)
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        presenter.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        presenter.onConfigurationChanged(newConfig)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        presenter.onSaveInstanceState(outState)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        presenter.onRestoreInstanceState(savedInstanceState)
    }

    override fun <Component> getComponent(componentClass: Class<Component>): Component? {
        return if (componentClass.isAssignableFrom(EventBus::class.java)) {
            EventBus.getDefault() as Component
        } else null
    }

}
